app/src/main/java/de/westnordost/streetcomplete/view/image_select/GroupedImageSelectAdapter.kt

```diff
@@ -51,36 +68,42 @@ class GroupedImageSelectAdapter<T>(val gridLayoutManager: GridLayoutManager) :
 
     override fun onBindViewHolder(holder: ItemViewHolder, position: Int) {
         holder.bind(_items[position])
-        holder.isSelected = selectedItem?.let { _items.indexOf(it) == position } == true
-        holder.isGroupExpanded = getGroup(selectedIndex) == position
+        holder.isSelected = selectedIndex == position
+        holder.isGroupExpanded = getGroupIndex(selectedIndex) == position
     }
 
     private fun toggle(index: Int) {
-        val prevSelectedItem = selectedItem
-        if (selectedItem == null || prevSelectedItem !== _items[index]) {
+        val previousSelectedItem = selectedItem
+        val previousSelectedItemGroup = selectedItemGroup
+
+        val foundGroupIndex = getGroupIndex(index)
+        if (foundGroupIndex == -1) {
+            selectedItemGroup = null
+        } else {
+            selectedItemGroup = _items[foundGroupIndex]
+        }
+        if (selectedItem == null || previousSelectedItem !== _items[index]) {
             selectedItem = _items[index]
         } else {
             selectedItem = null
         }
 
-        val selectedItem = selectedItem
-        if (prevSelectedItem != null) {
-            val prevSelectedIndex = _items.indexOf(prevSelectedItem)
-            notifyItemChanged(prevSelectedIndex)
-
-            val previousGroupIndex = getGroup(prevSelectedIndex)
-            if (previousGroupIndex != -1) {
-                if (selectedItem == null || previousGroupIndex != getGroup(_items.indexOf(selectedItem))) {
+        if (previousSelectedItem != null) {
+            val previousItemIndex = indexOfItemGivenGroupMembership(previousSelectedItem, previousSelectedItemGroup)
+            val previousGroupIndex = getGroupIndex(previousItemIndex)
+            notifyItemChanged(previousItemIndex)
+            if (previousSelectedItemGroup != null) {
+                if (selectedItem == null || previousSelectedItemGroup != selectedItemGroup) {
                     retractGroup(previousGroupIndex)
                 }
             }
         }
-        if (selectedItem != null) {
-            val selectedIndex = _items.indexOf(selectedItem)
+        val copyOfSelectedItem = selectedItem
```

Comment:
"Copy" may be misleading, because it is not a copy. It points to the same object in memory.